### PR TITLE
feat: store partial page-data results in lmdb

### PR DESCRIPTION
**Pull request by pieh**
## Description

This uses `lmdb-store` backed cache instance for storing/retrieving page query results (to generate final `page-data` file once everything is ready) instead of writing temporary files and reading them out later.

## Benchmarks

Before change (LMDB, PQR, 5 workers):

```
run queries in workers - 359.396s - 25848/25848 71.92/s
Writing page-data.json files to public directory - 77.187s - 25847/25847 334.86/s
Total for those activities: 436.583s
```

With this change (LMDB, PQR, 5 workers):

```
run queries in workers - 301.939s - 25848/25848 85.61/s
Writing page-data.json files to public directory - 79.732s - 25847/25847 324.17/s
Total for those activities: 381.671s
```

So for test site with 5 workers this yields ~12% speedup (for activities/steps that make use of page query results)

## Related Issues

[ch34445]

### CODE REVIEW

File: packages/gatsby/src/utils/page-data.ts [hunk 1]
```diff
@@ -67,6 +69,54 @@ export function pageDataExists(publicDir: string, pagePath: string): boolean {
   return fs.existsSync(getFilePath(publicDir, pagePath))
 }
 
+let lmdbPageQueryResultsCache: GatsbyCacheLmdb
+function getLMDBPageQueryResultsCache(): GatsbyCacheLmdb {
+  if (!lmdbPageQueryResultsCache) {
+    const GatsbyCacheLmdbImpl = require(`./cache-lmdb`).default
+    lmdbPageQueryResultsCache = new GatsbyCacheLmdbImpl({
```
Comment (LekoArts) [inline packages/gatsby/src/utils/page-data.ts:76]: I think it's missing an `.init()` here

Comment (vladar) [reply packages/gatsby/src/utils/page-data.ts:76]: Maybe we can just use `get-cache` utility? https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/utils/get-cache.ts

I know it's being used for plugins but it does pretty much the same thing? Maybe we can use some `internal-tmp-query-results` name for it?

Comment (pieh) [reply packages/gatsby/src/utils/page-data.ts:76]: > Maybe we can just use get-cache utility?

I don't think we want to use legacy cache implemenation for non-lmdb mode (as that would keep some page query results in memory), so using `get-cache` would look really similar with added extra abstraction? Not sure if worth it?

> I know it's being used for plugins but it does pretty much the same thing? Maybe we can use some internal-tmp-query-results name for it?

👍 for cache name suggestion

File: packages/gatsby/src/utils/page-data.ts [hunk 2]
```diff
@@ -67,6 +69,54 @@ export function pageDataExists(publicDir: string, pagePath: string): boolean {
...
+}
+
+export async function savePageQueryResult(
+  programDir: string,
+  pagePath: string,
+  result: IExecutionResult,
+  stringifiedResult: string
+): Promise<void> {
+  if (isLmdbStore()) {
+    await getLMDBPageQueryResultsCache().set(pagePath, result)
```
Comment (vladar) [inline packages/gatsby/src/utils/page-data.ts:90]: I am not sure if we should `await` it here. LMDB-store is designed for batched transactions, `await`-ing can potentially make it slower, but worth benchmarking I guess.

Comment (vladar) [reply packages/gatsby/src/utils/page-data.ts:90]: Also in my benchmark JSON-encoding results was also time-consuming. Would be great if we could switch to using query hash not based on results? Then we could save a bit more time here with `msgpackr` in lmdb.

Probably out of scope for this PR though.

Comment (pieh) [reply packages/gatsby/src/utils/page-data.ts]: Dropped await when storing and also switched to storing stringified results (with string encoding on db) in https://github.com/gatsbyjs/gatsby/pull/32431/commits/fa84ff96e34b68982307a3e99c5c85d2948b510d

File: packages/gatsby/src/utils/cache-lmdb.ts [hunk 3]
```diff
@@ -16,12 +16,20 @@ const cacheDbFile =
...
   private db: Database | undefined
+  private readonly encoding: DatabaseOptions["encoding"]
   public readonly name: string
   // Needed for plugins that want to write data to the cache directory
   public readonly directory: string
 
-  constructor({ name = `db` }: { name: string }) {
+  constructor({
+    name = `db`,
+    encoding = `msgpack`,
```
Comment (LekoArts) [inline packages/gatsby/src/utils/cache-lmdb.ts:26]: From https://github.com/gatsbyjs/gatsby/pull/32466

Should this be `json` instead?

```suggestion
    encoding = `json`,
```

**Review by LekoArts — APPROVED** [review]
👍

**Review by LekoArts — APPROVED** [review]
👍

**Comment by simonjoom** [comment]
Why you not let LmdbStore take care about page-data by default if the compilation speed improve at 12%? 
personaly i removed the test datastore.isLmdbStore() to use LMDB_STORE only for page-data
 
I don't use the env GATSBY_EXPERIMENTAL_LMDB_STORE for my project

**Comment by simonjoom** [comment]
I found an issue in your code from this pr

`async function savePageQueryResult(programDir, pagePath, stringifiedResult) {
   if ((0, _datastore.isLmdbStore)()) {
    //isssue here no await
    //savePageQueryResultsPromise = getLMDBPageQueryResultsCache().set(pagePath, stringifiedResult);
    await getLMDBPageQueryResultsCache().set(pagePath, stringifiedResult);
  } else {
    const pageQueryResultsPath = _path.default.join(programDir, `.cache`, `json`, `${pagePath.replace(/\//g, `_`)}.json`);

    await _fsExtra.default.outputFile(pageQueryResultsPath, stringifiedResult);
  } 
}`

**Pull request merged**